storage-proofs/src/hasher/poseidon.rs

```diff
@@ -0,0 +1,453 @@
+// https://github.com/briansmith/ring/blob/abb3fdfc08562f3f02e95fb551604a871fd4195e/src/polyfill.rs#L93-L110
+#[inline(always)]
+#[allow(clippy::needless_lifetimes)]
+fn as_ref<'a>(src: &'a [u64; 4]) -> &'a [u8] {
+    unsafe {
+        std::slice::from_raw_parts(
+            src.as_ptr() as *const u8,
+            src.len() * std::mem::size_of::<u64>(),
+        )
+    }
+}
+
+impl Domain for PoseidonDomain {
+    // QUESTION: When, if ever, should serialize and into_bytes return different results?
+    // The definitions here at least are equivalent.
+    // I'm taking one step toward resolving this by formalizing that equivalence while copying this base implementation from perdersen.rs. -porcuquine.
+    fn serialize(&self) -> Vec<u8> {
+        self.into_bytes()
+    }
+
+    fn into_bytes(&self) -> Vec<u8> {
+        let mut out = Vec::with_capacity(PoseidonDomain::byte_len());
+        self.0.write_le(&mut out).unwrap();
+
+        out
+    }
+
+    fn try_from_bytes(raw: &[u8]) -> Result<Self> {
+        ensure!(raw.len() == PoseidonDomain::byte_len(), Error::BadFrBytes);
+        let mut res: FrRepr = Default::default();
+        res.read_le(raw)?;
+
+        Ok(PoseidonDomain(res))
+    }
+
+    fn write_bytes(&self, dest: &mut [u8]) -> Result<()> {
+        self.0.write_le(dest)?;
+        Ok(())
+    }
+
+    fn random<R: rand::RngCore>(rng: &mut R) -> Self {
+        // generating an Fr and converting it, to ensure we stay in the field
+        Fr::random(rng).into()
+    }
+}
+
+impl Element for PoseidonDomain {
+    fn byte_len() -> usize {
+        32
+    }
+
+    fn from_slice(bytes: &[u8]) -> Self {
+        match PoseidonDomain::try_from_bytes(bytes) {
+            Ok(res) => res,
+            Err(err) => panic!(err),
+        }
+    }
+
+    fn copy_to_slice(&self, bytes: &mut [u8]) {
+        bytes.copy_from_slice(&self.into_bytes());
+    }
+}
+
+impl StdHasher for PoseidonFunction {
+    #[inline]
+    fn write(&mut self, msg: &[u8]) {
+        self.0 = Fr::from_repr(shared_hash(msg).0).unwrap();
+    }
+
+    #[inline]
+    fn finish(&self) -> u64 {
+        unimplemented!()
+    }
+}
+
+fn shared_hash(data: &[u8]) -> PoseidonDomain {
+    // FIXME: We shouldn't unwrap here, but doing otherwise will require an interface change.
+    // We could truncate so `bytes_into_frs` cannot fail, then ensure `data` is always `fr_safe`.
+    let preimage = data
+        .chunks(32)
+        .map(|ref chunk| {
+            <Bls12 as ff::ScalarEngine>::Fr::from_repr(PoseidonDomain::from_slice(chunk).0).unwrap()
+        })
+        .collect::<Vec<_>>();
+
+    shared_hash_frs(&preimage)
+}
+
+fn shared_hash_frs(preimage: &[<Bls12 as ff::ScalarEngine>::Fr]) -> PoseidonDomain {
+    let mut p = Poseidon::new_with_preimage(&preimage, &*POSEIDON_CONSTANTS);
+    let fr: <Bls12 as ScalarEngine>::Fr = p.hash_in_mode(HashMode::OptimizedStatic);
```

Comment:
It might be better to just call `p.hash()` here, as that is the default. That should probably be made the only public method, so don't have to continue to support that interface — which should probably be internal to the library.

Reply:
Ah for some reason I thought that was not the default, will switch

Reply:
fixed